HIP/hip-513.md

```diff
@@ -0,0 +1,342 @@
+
+    /**
+     * The type of this action.
+     */
+    ContractActionType call_type = 1;
+
+    /**
+     * Only the first action can come from an account, the rest will come from contracts.  Because of DELEGATECALL
+     * and CALLCODE the caller of actions whose parent is an account may also be an account.
+     */
+    oneof caller {
+        /**
+         * If the caller was a regular account, the AccountID.
+         */
+        AccountID calling_account = 2;
+
+        /**
+         * If the caller was a smart contract account, the ContractID.
+         */
+        ContractID calling_contract = 3;
+    }
+
+    /**
+     * Gas given to this action, the upper limit of gas this action can spend.
+     */
+    int64 gas = 4;
+
+    /**
+     * Bytes passed in as input data to this action.
+     */
+    bytes input = 5;
+
+    /**
+     * Who this action is directed to.
+     */
+    oneof recipient {
+        /**
+         * The AccountID of the recipient if the recipient is an account. Only HBars will be transferred, no other side
+         * effects should be expected.
+         */
+        AccountID recipient_account = 6;
+
+        /**
+         * The ContractID of the recipient if the recipient is a smart contract.
+         */
+        ContractID recipient_contract = 7;
+
+        /**
+         * If the action was directed at an invalid solidity address, what that address was.
+         */
+        bytes invalid_solidity_address = 8;
+    }
+
+    /**
+     * The value (in tinybars) that is associated with this action.
+     */
+    int64 value = 9;
+
+    /**
+     * The actual gas spent by this action.
+     */
+    uint64 gas_used = 10;
+
+    /**
+     * The result data of the action.
+     */
+    oneof resultData {
+
+        /**
+         * If successful, the output bytes of the action.
+         */
+        bytes output = 11;
+
+        /**
+         * The contract itself caused the transaction to fail via the `REVERT` operation
+         */
+        bytes revert_reason = 12;
+
+        /**
+         * The transaction itself failed without an explicit `REVERT`
+         */
+        bytes error = 13;
+    }
+
+    /**
+     * The nesting depth of this call. The original action is at depth=0.
+     */
+    int32 call_depth = 14;
+}
+```
+
+For every smart contract transaction a `ContractAction` record should be added to the sidecar record.
+
+### Add ContractBytecode sidecar record type
+
+Add and populate a new sidecar record of type `ContractBytecodes`. It will
+have a single field - repeated `ContractBytecode` with two fields - `bytecode` 
+and `deployed_bytecode`. Fields will allow tracking smart contracts.
+
+```
+message ContractBytecodes {
+    repeated ContractBytecode contract_bytecodes = 1;
+}
+
+message ContractBytecode {
+
+    /**
+    * Contract bytecode during deployment
+    */
+    bytes bytecode = 1;
+    
+    /**
+    * Contract bytecode after deployment
+    */
+    bytes deployed_bytecode = 2;
```

Comment:
consider `runtime_bytecode`, it is used. times in EIP specs vs 1 instance of "deployed bytecode"

Reply:
DONE